src/cursor/cur_layered.c

```diff
@@ -206,42 +206,229 @@ __clayered_close_cursors(WT_CURSOR_LAYERED *clayered, bool include_ingest)
+             * a follower and we never picked up a checkpoint, then no checkpoint has ever occurred
+             * on this Btree. Everything we need will be satisfied by the ingest table until the
+             * next checkpoint is picked up. So technically, opening this (empty) stable table is
+             * wasteful, but it's a corner case, it will be resolved at the next checkpoint, and it
+             * keeps the code easy.
+             */
+            cfg[2] = "readonly,checkpoint_use_history=false";
+            F_SET(clayered, WT_CLAYERED_STABLE_NO_CKPT);
+        } else {
+            /*
+             * Use a URI with a "/<checkpoint name> suffix. This is interpreted as reading from the
+             * stable checkpoint, but without it being a traditional checkpoint cursor.
+             */
+            WT_RET(__wt_snprintf(stable_uri_buf, sizeof(stable_uri_buf), "%s/%s",
+              layered->stable_uri, checkpoint_name));
+            stable_uri = stable_uri_buf;
+            cfg[2] = "readonly";
+        }
+    }
+    ret = __wt_open_cursor(session, stable_uri, &clayered->iface, cfg, &clayered->stable_cursor);
+
+    if (ret == ENOENT && !leader) {
+        /*
+         * This is fine, we may not have seen a checkpoint with this table yet. The open will be
+         * deferred.
+         */
+        ret = 0;
+    } else if (ret == WT_NOTFOUND)
+        WT_RET(
+          __wt_panic(session, WT_PANIC, "Layered table could not access stable table on leader"));
+
+    if (clayered->stable_cursor != NULL) {
+        F_SET(clayered->stable_cursor, WT_CURSTD_OVERWRITE | WT_CURSTD_RAW);
+
+        /* Layered cursor is not compatible with cursor_copy config. */
+        F_CLR(clayered->stable_cursor, WT_CURSTD_DEBUG_COPY_KEY | WT_CURSTD_DEBUG_COPY_VALUE);
+    }
+
+    return (ret);
+}
+
 /*
  * __clayered_adjust_state --
  *     Update the state of the cursor to match the state of the disaggregated system. In particular,
- *     if the system has changed in a way that makes constituent cursors out of date, close them.
- *     They will be opened later as needed.
+ *     if the system has changed in a way that makes constituent cursors out of date, either reopen
+ *     them or close them, and let them be opened later as needed.
  */
 static int
-__clayered_adjust_state(WT_CURSOR_LAYERED *clayered, bool *state_updated)
+__clayered_adjust_state(WT_CURSOR_LAYERED *clayered, bool iteration, bool *state_updated)
 {
     WT_CONNECTION_IMPL *conn;
-    uint64_t current_checkpoint_id;
-    bool current_leader;
+    WT_CURSOR *old_stable;
+    WT_SESSION_IMPL *session;
+    WT_TXN_SHARED *txn_shared;
+    uint64_t current_checkpoint_id, snapshot_gen;
+    bool change_ingest, change_stable, current_leader;
 
-    conn = S2C(CUR2S(clayered));
-    current_leader = conn->layered_table_manager.leader;
     *state_updated = false;
-
+    session = CUR2S(clayered);
+    conn = S2C(session);
+    current_leader = conn->layered_table_manager.leader;
```

Comment:
nitpick: I think we can only change leadership when no transaction is running so it is safe to read it this way. But it is better to use a atomic relaxed read to read it as leader is a shared variable.

Reply:
I'm not familiar with "atomic relaxed read". Are we talking about this?
```
        WT_ACQUIRE_READ(current_leader, conn->layered_table_manager.leader);
```
I'm in favor of doing things the safer way, though I don't see other places in the code that get the global "leader" variable, except as a naked read.

Reply:
I mean we can use __wt_atomic_load8. I understand why are not using that for now. So we can ignore it for now. However, we do need to clean it up some time in the future otherwise TSAN build will fail.

Reply:
Got it, thanks!